gemini/use-cases/retrieval-augmented-generation/qna_using_query_routing/utils/intent_routing.py

```diff
@@ -0,0 +1,413 @@
+
+        Returns:
+            list: A list of programming languages extracted from the query (potentially empty).
+        """
+
+        response = self.model.generate_content(
+            f"""
+            You are strict programming languages extractor.
+
+            Instructions:
+            1. Extract only programming languages from message.
+            2. Don't return any other languages other than programming.
+            3. return [] if no programming lanuage in mentioned in message.
+            4. {enabled_programming_language} these are the programming languages.
+
+            Examples:
+            write a code for fibonacci series in C++? : ["C++"]
+            write a code using C# to generate palindrome series : ["C#"]
+            using python, write a sample application code to create endpoint? : ["Python"]
+            what are classes in Java? : ["Java"]
+            write a code for reverse a string : []
+            what are data types? : []
+
+            What are the programming languages mentioned in below message?
+            MESSAGE:{text}
+            programming languages:""",  # pylint: disable=C0301:line-too-long
+            generation_config=self.genai_qna_parameters,
+        )
+
+        programming_lang = response.text
+        program_lang_in_query = []
+        if programming_lang:
+            try:
+                programming_lang = programming_lang.replace("'", '"')
+                programming_lang = json.loads(programming_lang)
+                program_lang_in_query = [
+                    x.lower().replace(" ", "").strip() for x in programming_lang
+                ]
+            except Exception:  # pylint: disable=W0718,W0703
+                self.logger.info("Error while extracting programming language.")
+        return program_lang_in_query
+
+    def check_programming_language_in_query(
+        self, text: str
+    ) -> tuple[List[str], set[str]]:
+        """
+        Identifies supported programming languages mentioned in a user's query.
+
+        Args:
+            text (str): The user's query.
+
+        Returns:
+            tuple:
+                * list: All programming languages found in the query.
+                * set: Programming languages in the query that are supported by the assistant.
+        """
+
+        enabled_programming_language = self.config["default"][
+            "enabled_programming_language"
+        ]
+        enabled_programming_language_list = enabled_programming_language.split(",")
+        enabled_programming_language_list = [
+            x.lower().replace(" ", "").strip()
+            for x in enabled_programming_language_list
+        ]
+        program_lang_in_query = self.get_programming_lanuage_from_query(
+            text, enabled_programming_language_list
+        )
+        allowed_language_in_query = set(enabled_programming_language_list).intersection(
+            set(program_lang_in_query)
+        )
```

Comment:
This could probably just be added to the end of the `get_programming_language_from_query` method since it's just another function call.

Or add it to the function that calls this one, so you don't have to return the tuple.

Reply:
updated the code block from the function, reducing the number of functions.